### PR TITLE
[runtime] Use MonoError for mono_object_register_finalizer

### PR DESCRIPTION
**Pull request by ludovic-henry**

### CODE REVIEW

File: mono/utils/mono-error.c [hunk 1]
```diff
@@ -451,6 +451,18 @@ mono_error_set_argument (MonoError *oerror, const char *argument, const char *ms
 }
 
 void
+mono_error_set_argument_null (MonoError *oerror, const char *argument, const char *msg_format, ...)
+{
+	MonoErrorInternal *error = (MonoErrorInternal*)oerror;
+	mono_error_prepare (error);
+
+	error->error_code = MONO_ERROR_ARGUMENT_NULL;
+	error->type_name = argument; /*use the first available string slot*/
```
Comment (kumpera) [inline mono/utils/mono-error.c]: On my recent cleanup of mono-error I fixed that mistake. We now have a field to store those so we don't break mono_errror_get_message. See https://github.com/mono/mono/commit/4df79f93834a018962eeea4426134ef0716a835f

File: mono/utils/mono-error.c [hunk 2]
```diff
@@ -611,6 +623,10 @@ mono_error_prepare_exception (MonoError *oerror, MonoError *error_out)
 		exception = mono_get_exception_argument (error->type_name, mono_internal_error_get_message (error));
 		break;
 
+	case MONO_ERROR_ARGUMENT_NULL:
```
Comment (kumpera) [inline mono/utils/mono-error.c]: Unfortunately, new error kinds require us to support the conversion to loader error for now.

Meaning you need to change mono_error_set_from_loader_error and mono_loader_set_error_from_mono_error too.

Ideally this would be a separate commit from the one that introduced MonoError into mono_object_register_finalizer.

Comment (ludovic-henry) [inline mono/utils/mono-error.c]: I based this code on how `MONO_ERROR_ARGUMENT` is handled, and it is not handled neither in `mono_error_set_from_loader_error` neither in `mono_loader_set_error_from_mono_error`. So is it something we need to add for both `MONO_ERROR_ARGUMENT` and `MONO_ERROR_ARGUMENT_NULL`, or none of them?

**Comment by kumpera** [comment]
The MonoError change conflicts with @lambdageek change to the MonoError enum.

**Comment by kumpera** [comment]
I think it's weird to have a function that returns void and yet could fail.

It should at least return a boolean telling whether it worked or not.

File: mono/utils/mono-error.h [hunk 3]
```diff
@@ -24,6 +24,7 @@ enum {
 	MONO_ERROR_BAD_IMAGE = 5,
 	MONO_ERROR_OUT_OF_MEMORY = 6,
 	MONO_ERROR_ARGUMENT = 7,
+	MONO_ERROR_ARGUMENT_NULL = 10,
```
Comment (kumpera) [inline mono/utils/mono-error.h]: This number conflicts with the recent addition to MonoError.

**Comment by lambdageek** [comment]
I'll update #2516 to use `11` for its enum value.

**Comment by ludovic-henry** [comment]
@aleksey no need to change yours, I have to update this PR anyway

**Comment by ludovic-henry** [comment]
build

File: mono/utils/mono-error.c [hunk 4]
```diff
@@ -604,6 +616,10 @@ mono_error_prepare_exception (MonoError *oerror, MonoError *error_out)
 		exception = mono_get_exception_argument (error->first_argument, error->full_message);
 		break;
 
+	case MONO_ERROR_ARGUMENT_NULL:
+		exception = mono_get_exception_argument_null (error->type_name);
```
Comment (lambdageek) [inline mono/utils/mono-error.c]: Should be `error->first_argument`

**Comment by ludovic-henry** [comment]
So out of date, it's easier to close it and restart.
